includes/CreditCardTokenControl.php

```diff
@@ -76,125 +86,130 @@ function createCreditCardDB()
-            if($paymentMethods[$orderPaymentMethod]->settings['enabled']=='yes' && $paymentMethods[$orderPaymentMethod]->settings['token_control']=='yes'){
-                return '<form method="post" action="">
+	require_once ABSPATH . '/wp-admin/includes/upgrade.php';
+	dbDelta( $sql );
+	add_option( '$creditCardDBVersion', $creditCardDBVersion );
+}
+
+/**
+ * Add Dropdown list of saved credit cards in description of payment options on checkout page
+ *
+ * @param string $description
+ * @param string $payment_id
+ *
+ * @return string
+ */
+function creditCardCustomDescription( $description, $payment_id ) {
+	$gateways = WC()->payment_gateways->get_available_payment_gateways();
+
+	$checkout = new WC_Checkout();
+	if ( $gateways ) {
+		foreach ( $gateways as $gateway ) {
+
+			if ( $gateway->enabled === 'yes' && isset( $gateway->settings['token_control'] ) && $gateway->settings['token_control'] === 'yes' ) {
+
+				if ( $gateway->id === $payment_id && is_user_logged_in() ) {
+					ob_start();
+					global $wpdb;
+
+					$userID  = get_current_user_id();
+					$results = $wpdb->get_results( "SELECT * FROM {$wpdb->prefix}altapayCreditCardDetails WHERE userID='$userID'" );
+
+					$creditCard[] = _( 'Select a saved credit card' );
+					foreach ( $results as $result ) {
+						$creditCard[ $result->creditCardNumber ] = $result->creditCardNumber . ' (' . $result->cardExpiryDate . ')';
+					}
+
+					woocommerce_form_field(
+						'savedCreditCard',
+						array(
+							'type'    => 'select',
+							'class'   => array( 'wps-drop' ),
+							'options' => $creditCard,
+							'default' => '',
+						),
+						$checkout->get_value( 'savedCreditCard' )
+					);
+					$description .= ob_get_clean(); // Append buffered content
+				}
+			}
+		}
+	}
+
+	return $description;
+}
+
+
+/**
+ * Display Saved Credit Card Details button after successful order
+ *
+ * @param string   $text
+ * @param WC_Order $order
+ *
+ * @return string
+ */
+function filterSaveCreditCardDetailsButton( $text, $order ) {
+	global $wpdb;
+
+	// Get payment methods
+	$paymentMethods = WC()->payment_gateways->get_available_payment_gateways();
+
+	$orderMeta          = get_metadata( 'post', $order->id );
+	$cardNo             = $orderMeta['_cardno'][0];
+	$ccToken            = $orderMeta['_credit_card_token'][0];
+	$ccBrand            = $orderMeta['_credit_card_brand'][0];
+	$ccExpiryDate       = $orderMeta['_credit_card_expiry_date'][0];
+	$orderPaymentMethod = $orderMeta['_payment_method'][0];
+	$buttonText         = _( 'Save your credit card for later use' );
+
+	$results = $wpdb->get_results( "SELECT * FROM {$wpdb->prefix}altapayCreditCardDetails WHERE userID='" . get_current_user_id() . "' and ccToken='$ccToken'" );
+
+	if ( array_key_exists( 'test', $_POST ) ) {
+		saveCreditCardDetails( $cardNo, $ccToken, $ccBrand, $ccExpiryDate );
+	}
+
+	if ( ! empty( $orderMeta['_cardno'][0] ) && empty( $results ) && is_user_logged_in() ) {
```

Comment:
empty() should only be used when the subject existence is unknown
```suggestion
	if ( ! empty( $orderMeta['_cardno'][0] ) && ! $results && is_user_logged_in() ) {
```
This is because empty() can hide typos in variable names to code analyzers. There is a lot of this in the code.